File: ios/Releaf/Data/Drive/PdfExporter.swift
/*
 * PdfExporter.swift
 *
 * Renders a `Page` to a single-page (or multi-page if it overflows)
 * letter-size PDF and writes it into `.documentDirectory`. Returns
 * the URL the page was written to so the caller can hand it to a
 * share sheet.
 *
 * The output is plain editorial: green eyebrow + serif title +
 * dateline, then the page's notes set as serif body, then a small
 * "captures summary" block at the bottom (counts per kind). It's
 * intentionally not a pixel-faithful render of the in-app screen —
 * users sharing PDFs of their pages want a readable doc, not a
 * screenshot. The visual language stays Releaf — same warm cream,
 * coral leaf, dot-grid suggestion — at print densities.
 *
 * Pure UIKit/CoreGraphics, no SwiftUI. iOS 16 floor; the renderer
 * APIs predate that comfortably.
 */

import Foundation
import UIKit

public enum PdfExporter {

    public static let pageWidth: CGFloat  = 612   // 8.5"
    public static let pageHeight: CGFloat = 792   // 11"
    public static let margin: CGFloat     = 56

    /// Render `page` to a fresh file under `.documentDirectory` and
    /// return the URL. Filename pattern: `page-{id}-{timestamp}.pdf`.
    public static func export(page: Page) throws -> URL {
        let renderer = UIGraphicsPDFRenderer(
            bounds: CGRect(x: 0, y: 0, width: pageWidth, height: pageHeight)
        )
        let data = renderer.pdfData { ctx in
            ctx.beginPage()
            var y = margin
            y = drawHeader(in: ctx.cgContext, page: page, top: y)
            y += 24
            y = drawNotes(in: ctx.cgContext, page: page, top: y, ctx: ctx)
            y += 24
            _ = drawCapturesSummary(in: ctx.cgContext, page: page, top: y)
        }

        let url = try documentsURL(for: page)
        try data.write(to: url, options: .atomic)
        return url
    }

    // MARK: - Sections

    private static func drawHeader(
        in cg: CGContext,
        page: Page,
        top: CGFloat
    ) -> CGFloat {
        var y = top
        let eyebrow = "RELEAF · PAGE"
        let eyebrowFont = UIFont.systemFont(ofSize: 9, weight: .medium)
        let eyebrowAttrs: [NSAttributedString.Key: Any] = [
            .font: eyebrowFont,
            .foregroundColor: themeGreenDeep,
            .kern: 1.6,
        ]
        let eyebrowSize = (eyebrow as NSString).size(withAttributes: eyebrowAttrs)
        (eyebrow as NSString).draw(at: CGPoint(x: margin, y: y), withAttributes: eyebrowAttrs)

        // Coral leaf glyph next to eyebrow
        let glyph = leafPath(at: CGPoint(x: margin + eyebrowSize.width + 6, y: y + 1),
                             height: eyebrowSize.height - 2)
        cg.setFillColor(coral.cgColor)
        cg.addPath(glyph)
        cg.fillPath()

        y += eyebrowSize.height + 8

        // Title
        let titleFont = UIFont.systemFont(ofSize: 28, weight: .regular)
        let titleAttrs: [NSAttributedString.Key: Any] = [
            .font: serifVariant(of: titleFont),
            .foregroundColor: ink,
        ]
        let titleRect = CGRect(x: margin, y: y, width: pageWidth - 2 * margin, height: 80)
        (page.title as NSString).draw(in: titleRect, withAttributes: titleAttrs)
        y += titleFont.lineHeight + 4

        // Dateline + notebook context
        if let captured = page.capturedOn {
            let metaFont = serifVariant(of: UIFont.italicSystemFont(ofSize: 13))
            let metaAttrs: [NSAttributedString.Key: Any] = [
                .font: metaFont,
                .foregroundColor: inkSoft,
            ]
            (captured as NSString).draw(at: CGPoint(x: margin, y: y), withAttributes: metaAttrs)
            y += metaFont.lineHeight + 6
        }

        // Hairline rule
        cg.setStrokeColor(borderRule.cgColor)
        cg.setLineWidth(0.5)
        cg.move(to: CGPoint(x: margin, y: y + 8))
        cg.addLine(to: CGPoint(x: pageWidth - margin, y: y + 8))
        cg.strokePath()

        return y + 16
    }

    private static func drawNotes(
        in cg: CGContext,
        page: Page,
        top: CGFloat,
        ctx: UIGraphicsPDFRendererContext
    ) -> CGFloat {
        var y = top
        guard !page.notes.isEmpty else {
            let muted = UIFont.italicSystemFont(ofSize: 13)
            ("(no notes)" as NSString).draw(
                at: CGPoint(x: margin, y: y),
                withAttributes: [.font: muted, .foregroundColor: inkSoft]
            )
            return y + muted.lineHeight
        }

        let bodyFont = serifVariant(of: UIFont.systemFont(ofSize: 13))
        let attrs: [NSAttributedString.Key: Any] = [
            .font: bodyFont,
            .foregroundColor: ink,
        ]
        for (idx, note) in page.notes.enumerated() {
            let bullet = "—"
            (bullet as NSString).draw(
                at: CGPoint(x: margin, y: y),
                withAttributes: [.font: bodyFont, .foregroundColor: themeGreenDeep]
            )
            let textRect = CGRect(
                x: margin + 16,
                y: y,
                width: pageWidth - 2 * margin - 16,
                height: pageHeight - margin - y
            )
            let body = NSString(string: note.body)
            let bounding = body.boundingRect(
                with: CGSize(width: textRect.width, height: .greatestFiniteMagnitude),
                options: [.usesLineFragmentOrigin, .usesFontLeading],
                attributes: attrs,
                context: nil
            )
            // Page-break safety — if this paragraph won't fit, end the page.
            if y + bounding.height > pageHeight - margin {
                ctx.beginPage()
                y = margin
            }
            body.draw(in: textRect, withAttributes: attrs)
            y += bounding.height + (idx == page.notes.count - 1 ? 0 : 12)
        }
        return y
    }

    private static func drawCapturesSummary(
        in cg: CGContext,
        page: Page,
        top: CGFloat
    ) -> CGFloat {
        let counts = page.counts
        let pieces: [String] = [
            counts.photos > 0           ? "\(counts.photos) photo\(counts.photos == 1 ? "" : "s")" : nil,
            counts.scannedDocuments > 0 ? "\(counts.scannedDocuments) scan\(counts.scannedDocuments == 1 ? "" : "s")" : nil,
            counts.voiceNotes > 0       ? "\(counts.voiceNotes) voice note\(counts.voiceNotes == 1 ? "" : "s")" : nil,
            counts.todoItems > 0        ? "\(counts.todoItems) to-do\(counts.todoItems == 1 ? "" : "s")" : nil,
            counts.contacts > 0         ? "\(counts.contacts) contact\(counts.contacts == 1 ? "" : "s")" : nil,
            counts.locations > 0        ? "\(counts.locations) place\(counts.locations == 1 ? "" : "s")" : nil,
        ].compactMap { $0 }

        guard !pieces.isEmpty else { return top }

        let eyebrow = "CAPTURES"
        let eyebrowFont = UIFont.systemFont(ofSize: 9, weight: .medium)
        let eyebrowAttrs: [NSAttributedString.Key: Any] = [
            .font: eyebrowFont,
            .foregroundColor: inkSoft,
            .kern: 1.6,
        ]
        var y = top
        (eyebrow as NSString).draw(at: CGPoint(x: margin, y: y), withAttributes: eyebrowAttrs)
        y += eyebrowFont.lineHeight + 4

        let summary = pieces.joined(separator: " · ")
        let summaryFont = serifVariant(of: UIFont.systemFont(ofSize: 13))
        (summary as NSString).draw(
            at: CGPoint(x: margin, y: y),
            withAttributes: [.font: summaryFont, .foregroundColor: ink]
        )
        return y + summaryFont.lineHeight
    }

    // MARK: - Helpers

    private static func documentsURL(for page: Page) throws -> URL {
        let docs = try FileManager.default.url(
            for: .documentDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let safeId = page.id.replacingOccurrences(of: "/", with: "-")
        let stamp = Int(Date().timeIntervalSince1970)
        return docs.appendingPathComponent("page-\(safeId)-\(stamp).pdf")
    }

    /// Returns a serif version of `font` at the same point size +
    /// weight. Falls back to the original font if the system can't
    /// find a serif variant (rare; iOS ships New York / Times).
    private static func serifVariant(of font: UIFont) -> UIFont {
        let descriptor = font.fontDescriptor.withDesign(.serif) ?? font.fontDescriptor
        return UIFont(descriptor: descriptor, size: font.pointSize)
    }

    /// A small leaf-shaped path centered on `at` with the given
    /// height. Used as the inline glyph next to the eyebrow.
    private static func leafPath(at origin: CGPoint, height: CGFloat) -> CGPath {
        let path = CGMutablePath()
        let w = height * 0.7
        path.move(to: CGPoint(x: origin.x + w / 2, y: origin.y))
        path.addQuadCurve(
            to: CGPoint(x: origin.x + w / 2, y: origin.y + height),
            control: CGPoint(x: origin.x + w * 1.15, y: origin.y + height * 0.45)
        )
        path.addQuadCurve(
            to: CGPoint(x: origin.x + w / 2, y: origin.y),
            control: CGPoint(x: origin.x - w * 0.15, y: origin.y + height * 0.45)
        )
        return path
    }

    // MARK: - Tokens (light-mode hex literals)
    //
    // The PDF doesn't run through `dynamicColor` — it's a static
    // file shared off-device. We pin to the light-mode Releaf
    // palette so what the user sees in their PDF reader matches the
    // brand on a printed page.

    private static let ink           = UIColor(red: 0x46/255, green: 0x3C/255, blue: 0x31/255, alpha: 1)
    private static let inkSoft       = UIColor(red: 0x5F/255, green: 0x52/255, blue: 0x45/255, alpha: 1)
    private static let coral         = UIColor(red: 0xE0/255, green: 0x78/255, blue: 0x56/255, alpha: 1)
    private static let themeGreenDeep = UIColor(red: 0x5B/255, green: 0x8C/255, blue: 0x52/255, alpha: 1)
    private static let borderRule    = UIColor(red: 0x50/255, green: 0x3E/255, blue: 0x2D/255, alpha: 0.24)
}

File: apps/releaf/android/app/src/main/java/app/releaf/mobile/data/drive/PdfExporter.kt
/*
 * PdfExporter.kt
 *
 * Renders a `Page` to a single-page (or multi-page if it overflows)
 * letter-size PDF and writes it under `cacheDir/exports/` so the
 * file is share-target reachable via FileProvider. Returns the
 * Uri the page was written to.
 *
 * The output is plain editorial: green eyebrow + serif title +
 * dateline, then the page's notes set as serif body, then a small
 * captures-summary block at the bottom. Mirrors PdfExporter.swift
 * — same layout, same warm tokens, same fonts (system serif via
 * Typeface.SERIF since we don't ship Newsreader on Android yet).
 */

package app.releaf.mobile.data.drive

import android.content.Context
import android.graphics.Color as AColor
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.RectF
import android.graphics.Typeface
import android.graphics.pdf.PdfDocument
import androidx.core.content.FileProvider
import app.releaf.mobile.data.domain.Page
import java.io.File
import java.io.FileOutputStream
import java.time.Instant

object PdfExporter {

    private const val PAGE_WIDTH  = 612      // 8.5"
    private const val PAGE_HEIGHT = 792      // 11"
    private const val MARGIN      = 56f

    private val ink            = AColor.rgb(0x46, 0x3C, 0x31)
    private val inkSoft        = AColor.rgb(0x5F, 0x52, 0x45)
    private val coral          = AColor.rgb(0xE0, 0x78, 0x56)
    private val themeGreenDeep = AColor.rgb(0x5B, 0x8C, 0x52)
    private val borderRule     = AColor.argb(0x3D, 0x50, 0x3E, 0x2D) // ~24%

    /** Render [page] to a fresh file under `cacheDir/exports/` and
     *  return a content:// Uri reachable via the app's FileProvider.
     *  Filename pattern: `page-{id}-{timestamp}.pdf`. */
    fun export(context: Context, page: Page): android.net.Uri {
        val document = PdfDocument()
        val pageInfo = PdfDocument.PageInfo.Builder(PAGE_WIDTH, PAGE_HEIGHT, 1).create()
        val pdfPage  = document.startPage(pageInfo)
        val canvas   = pdfPage.canvas

        var y = MARGIN
        y = drawHeader(canvas, page, y)
        y += 24f
        y = drawNotes(canvas, page, y)
        y += 24f
        drawCapturesSummary(canvas, page, y)

        document.finishPage(pdfPage)

        val file = exportFile(context, page)
        FileOutputStream(file).use { document.writeTo(it) }
        document.close()

        return FileProvider.getUriForFile(
            context,
            "${context.packageName}.fileprovider",
            file,
        )
    }

    // ---------- Sections ----------

    private fun drawHeader(canvas: android.graphics.Canvas, page: Page, top: Float): Float {
        var y = top

        val eyebrow = "RELEAF · PAGE"
        val eyebrowPaint = Paint().apply {
            isAntiAlias = true
            color       = themeGreenDeep
            textSize    = 9f
            typeface    = Typeface.create(Typeface.SANS_SERIF, Typeface.BOLD)
            letterSpacing = 0.16f
        }
        val eyebrowMetrics = eyebrowPaint.fontMetrics
        canvas.drawText(eyebrow, MARGIN, y - eyebrowMetrics.ascent, eyebrowPaint)
        val eyebrowWidth = eyebrowPaint.measureText(eyebrow)

        // Coral leaf glyph next to eyebrow
        val glyphHeight = 9f
        drawLeafGlyph(canvas, MARGIN + eyebrowWidth + 6f, y + 1f, glyphHeight)

        y += eyebrowPaint.fontMetrics.run { descent - ascent } + 6f

        // Title
        val titlePaint = Paint().apply {
            isAntiAlias = true
            color       = ink
            textSize    = 28f
            typeface    = Typeface.SERIF
        }
        canvas.drawText(page.title, MARGIN, y - titlePaint.fontMetrics.ascent, titlePaint)
        y += titlePaint.fontMetrics.run { descent - ascent } + 4f

        // Dateline
        page.capturedOn?.let { captured ->
            val metaPaint = Paint().apply {
                isAntiAlias = true
                color       = inkSoft
                textSize    = 13f
                typeface    = Typeface.create(Typeface.SERIF, Typeface.ITALIC)
            }
            canvas.drawText(captured, MARGIN, y - metaPaint.fontMetrics.ascent, metaPaint)
            y += metaPaint.fontMetrics.run { descent - ascent } + 6f
        }

        // Hairline rule
        val rulePaint = Paint().apply {
            isAntiAlias = true
            color       = borderRule
            strokeWidth = 0.5f
        }
        canvas.drawLine(MARGIN, y + 8f, PAGE_WIDTH - MARGIN, y + 8f, rulePaint)

        return y + 16f
    }

    private fun drawNotes(canvas: android.graphics.Canvas, page: Page, top: Float): Float {
        var y = top

        if (page.notes.isEmpty()) {
            val mutedPaint = Paint().apply {
                isAntiAlias = true
                color       = inkSoft
                textSize    = 13f
                typeface    = Typeface.create(Typeface.SERIF, Typeface.ITALIC)
            }
            canvas.drawText("(no notes)", MARGIN, y - mutedPaint.fontMetrics.ascent, mutedPaint)
            return y + mutedPaint.fontMetrics.run { descent - ascent }
        }

        val bodyPaint = Paint().apply {
            isAntiAlias = true
            color       = ink
            textSize    = 13f
            typeface    = Typeface.SERIF
        }
        val bulletPaint = Paint(bodyPaint).apply { color = themeGreenDeep }
        val lineHeight = bodyPaint.fontMetrics.run { descent - ascent } + 2f
        val textWidth  = PAGE_WIDTH - 2 * MARGIN - 16f

        for ((idx, note) in page.notes.withIndex()) {
            val firstLineY = y - bodyPaint.fontMetrics.ascent
            canvas.drawText("—", MARGIN, firstLineY, bulletPaint)

            // Wrap the note body to the available width by scanning words.
            var line = StringBuilder()
            for (word in note.body.split(' ')) {
                val candidate = if (line.isEmpty()) word else "$line $word"
                if (bodyPaint.measureText(candidate) > textWidth) {
                    canvas.drawText(line.toString(), MARGIN + 16f, y - bodyPaint.fontMetrics.ascent, bodyPaint)
                    y += lineHeight
                    line = StringBuilder(word)
                } else {
                    line = StringBuilder(candidate)
                }
            }
            if (line.isNotEmpty()) {
                canvas.drawText(line.toString(), MARGIN + 16f, y - bodyPaint.fontMetrics.ascent, bodyPaint)
                y += lineHeight
            }
            if (idx != page.notes.lastIndex) y += 12f
        }
        return y
    }

    private fun drawCapturesSummary(canvas: android.graphics.Canvas, page: Page, top: Float): Float {
        val counts = page.counts
        val pieces = listOfNotNull(
            "${counts.photos} photo${pluralS(counts.photos)}".takeIf { counts.photos > 0 },
            "${counts.scannedDocuments} scan${pluralS(counts.scannedDocuments)}".takeIf { counts.scannedDocuments > 0 },
            "${counts.voiceNotes} voice note${pluralS(counts.voiceNotes)}".takeIf { counts.voiceNotes > 0 },
            "${counts.todoItems} to-do${pluralS(counts.todoItems)}".takeIf { counts.todoItems > 0 },
            "${counts.contacts} contact${pluralS(counts.contacts)}".takeIf { counts.contacts > 0 },
            "${counts.locations} place${pluralS(counts.locations)}".takeIf { counts.locations > 0 },
        )
        if (pieces.isEmpty()) return top

        var y = top
        val eyebrowPaint = Paint().apply {
            isAntiAlias = true
            color       = inkSoft
            textSize    = 9f
            typeface    = Typeface.create(Typeface.SANS_SERIF, Typeface.BOLD)
            letterSpacing = 0.16f
        }
        canvas.drawText("CAPTURES", MARGIN, y - eyebrowPaint.fontMetrics.ascent, eyebrowPaint)
        y += eyebrowPaint.fontMetrics.run { descent - ascent } + 4f

        val summaryPaint = Paint().apply {
            isAntiAlias = true
            color       = ink
            textSize    = 13f
            typeface    = Typeface.SERIF
        }
        canvas.drawText(pieces.joinToString(" · "), MARGIN, y - summaryPaint.fontMetrics.ascent, summaryPaint)
        return y + summaryPaint.fontMetrics.run { descent - ascent }
    }

    private fun pluralS(n: Int): String = if (n == 1) "" else "s"

    // ---------- Helpers ----------

    private fun exportFile(context: Context, page: Page): File {
        val dir = File(context.cacheDir, "exports").apply { mkdirs() }
        val safeId = page.id.replace('/', '-')
        val stamp  = Instant.now().epochSecond
        return File(dir, "page-$safeId-$stamp.pdf")
    }

    private fun drawLeafGlyph(
        canvas: android.graphics.Canvas,
        x: Float,
        y: Float,
        height: Float,
    ) {
        val paint = Paint().apply { isAntiAlias = true; color = coral }
        val w = height * 0.7f
        val path = android.graphics.Path().apply {
            moveTo(x + w / 2f, y)
            quadTo(x + w * 1.15f, y + height * 0.45f, x + w / 2f, y + height)
            quadTo(x - w * 0.15f, y + height * 0.45f, x + w / 2f, y)
            close()
        }
        canvas.drawPath(path, paint)
    }
}
